src/OverlayRendering.cpp

```diff
@@ -0,0 +1,271 @@
+	int32_t err, e2;
+
+    /*
+     * Clamp values to the extent of the bitmap.
+     * Probably extraneous as positions should be valid in the first place,
+     * else something is horribly wrong (with the caller).
+     * Clamping should be done using linear interpolation (lerp)... but this should be removed anyway...
+     * a*(1.0f-x)+(b*x), a, b : range, x : value.
+     */
+	if (x0 > r->w) { x0 = r->w; }
+	if (x1 > r->w) { x1 = r->w; }
+	if (y0 > r->h) { y0 = r->h; }
+	if (y1 > r->h) { y1 = r->h; }
+
+	dx  = abs((int32_t) x1 - (int32_t) x0);
+	sx  = (x0 > x1 ? 1 : -1);
+	dy  = abs((int32_t) y1 - (int32_t) y0);
+	sy  = (y0 > y1 ? 1 : -1);
+	err = (dx > dy ? dx : -dy) / 2;
+
+	for (;;)
+	{
+		if (x0 < r->w && y0 < r->h)
+		{
+			r->pixel[x0 + y0 * r->w] = colour;
+		}
+		if (x0 == x1 && y0 == y1)
+		{
+			break;
+		}
+		
+		e2 = err;
+		if (e2 > -dx)
+		{
+			err -= dy;
+			x0 = (uint32_t) ((int32_t) x0 - sx);
+		}
+		if (e2 < dy)
+		{
+			err += dx;
+			y0 = (uint32_t) ((int32_t) y0 - sy);
+		}
+	}
+}
+
+void
+OverlayLineRenderingModule::process_signal(core::Render2Signal& signal)
+{
+	for (const auto&& [ entity, line ] : AccessStorage<OverlayLine>::access_storage().each())
+	{
+        struct bitmap r;
+
+        r.pixel = signal.pixels;
+        r.w     = (uint32_t) signal.w;
+        r.h     = (uint32_t) signal.h;
+        r_line(&r, (uint32_t) line.start.x, (uint32_t) line.start.y, (uint32_t) line.end.x, (uint32_t) line.end.y, line.colour);
+	}
+}
+
+void
+OverlayTextRenderingModule::process_signal(core::Render2Signal& signal)
+{
+	for (const auto&& [ entity, textdata ] : AccessStorage<OverlayText>::access_storage().each())
+	{
+        struct bitmap r;
+
+        r.pixel = signal.pixels;
+        r.w     = (uint32_t) signal.w;
+        r.h     = (uint32_t) signal.h;
+        r_puts(&r, (uint32_t) textdata.position.x, (uint32_t) textdata.position.y, textdata.text.c_str());
+	}
+}
+
+void
+OverlayTestRenderingModule::process_signal(core::Render2Signal& signal)
+{
+	for (const auto&& [ entity, test ] : AccessStorage<OverlayTest>::access_storage().each())
+	{
+		uint32_t x, y;
+
+        switch (test.magic)
+        {
+        case 0:
+        for (y = 0; y < signal.h; y++)
+        {
+            for (x = 0; x < signal.w; x++)
+            {
+                uint32_t c = x^y;
+
+                signal.pixels[y * signal.w + x] = (c << 16) | (c << 8) | (x);
+            }
+        }
+        break;
+        case 1:
+        for (y = 0; y < signal.h; y++)
+        {
+            for (x = 0; x < signal.w; x++)
+            {
+                signal.pixels[y * signal.w + x] = (x << 16) | (x << 8) | (x);
+            }
+        }
+        default:
+        break;
+        }
+	}
+}
+
+/*
+ * Character width: 6, height: 11, total characters: 224, first character: ascii 32.
+ * Each line defines 2 characters, every character is made from 11 'unsigned char' bitmasks.
+ * Converted from BDF with bdf2c.
+ */
+uint8_t r_image_font_6_11[2464] =
```

Comment:
we _are_ including the bytes of a whole image, okay, just mark it very carefully as to be used for _testing_